cirq/contrib/acquaintance/permutation.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Sequence, Tuple, TypeVar
+
+from cirq import abc
+from cirq.ops import (
+        CompositeGate, Gate, QubitId, OP_TREE, SWAP,
+        flatten_op_tree, GateOperation, TextDiagrammable,
+        gate_features)
+
+LogicalIndex = TypeVar('LogicalIndex', int, QubitId)
+LogicalGates = Dict[Tuple[LogicalIndex, ...], Gate]
+LogicalMappingKey = TypeVar('LogicalMappingKey', bound=QubitId)
+LogicalMapping = Dict[LogicalMappingKey, LogicalIndex]
+
+class PermutationGate(Gate, TextDiagrammable, metaclass=abc.ABCMeta):
+    """Permutation gate."""
+
+    def __init__(self, swap_gate: Gate=SWAP) -> None:
+        self.swap_gate = swap_gate
+
+
+    @abc.abstractmethod
+    def permutation(self, qubit_count: int) -> Dict[int, int]:
+        """permutation = {i: s[i]} indicates that the i-th qubit is mapped to
+        the s[i]-th qubit."""
+        pass
+
+
+    def update_mapping(self, mapping: Dict[QubitId, LogicalIndex],
+                       keys: Sequence[QubitId]
+                       ) -> None:
+        n_elements = len(keys)
+        permutation = self.permutation(n_elements)
+        indices = tuple(permutation.keys())
+        new_keys = [keys[permutation[i]] for i in indices]
+        old_elements = [mapping[keys[i]] for i in indices]
+        mapping.update(zip(new_keys, old_elements))
+
+
+    @staticmethod
+    def validate_permutation(permutation: Dict[int, int],
+                             n_elements: int=None) -> None:
+        if not permutation:
+            return
+        if set(permutation.values()) != set(permutation):
```

Comment:
`if not (set(permutation.values()) == set(permutation) == set(range(n_elements)))`

Reply:
The intention is that the permutation dict doesn't need to "complete", in the sense that any index not explicitly permuted is left alone. For example, an empty permutation dict indicates the identity permutation.

Reply:
Ah, I see.